src/Npgsql.SourceGenerators/DatabaseInfoSourceGenerator.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+using Scriban;
+using Scriban.Runtime;
+
+namespace Npgsql.SourceGenerators;
+
+[Generator]
+sealed class DatabaseInfoSourceGenerator : ISourceGenerator
+{
+    public void Initialize(GeneratorInitializationContext context) { }
+
+    public void Execute(GeneratorExecutionContext context)
+    {
+        CatalogInfo? catalogs = null;
+        var hasCatalogJson = false;
+        foreach (var file in context.AdditionalFiles)
+        {
+            if (Path.GetFileName(file.Path) != "catalogs.json")
+                continue;
+
+            hasCatalogJson = true;
+            catalogs = JsonSerializer.Deserialize<CatalogInfo>(
```

Comment:
I'm not sure if it would be ok/reasonable to avoid doing this on every build, e.g. cache a hash of the input file across invocations and do nothing unless it changes (which is almost never will); otherwise this does have the potential to slow down our dev inner loop (analyzers and source generators are tough).

Seem my general comment above about avoiding this.

Reply:
Yeah, this is among the reasons why [incremental generators](https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md) have been introduced.
It turns out that they are pretty straightforward to use and solve this problem.

I've upgraded the generator to implement `IIncrementalGenerator` instead of `ISourceGenerator`.